shared/src/main/java/uk/gov/di/authentication/shared/lambda/UserContextAwareHandler.java

```diff
@@ -0,0 +1,133 @@
+                        configurationService.getEnvironment(),
+                        configurationService.getDynamoEndpointUri());
+    }
+
+    protected UserContextAwareHandler(
+            ConfigurationService configurationService,
+            SessionService sessionService,
+            ClientSessionService clientSessionService,
+            ClientService clientService) {
+        this.configurationService = configurationService;
+        this.sessionService = sessionService;
+        this.clientSessionService = clientSessionService;
+        this.clientService = clientService;
+    }
+
+    protected abstract APIGatewayProxyResponseEvent handleRequestDelegate(
+            APIGatewayProxyRequestEvent input, Context context);
+
+    protected Session getSession() {
+        return session;
+    }
+
+    protected ClientSession getClientSession() {
+        return clientSession;
+    }
+
+    protected ClientRegistry getClientRegistry() {
+        return clientRegistry;
+    }
+
+    protected Optional<APIGatewayProxyResponseEvent> initializeUserContext(
+            Map<String, String> headers) {
+
+        session = sessionService.getSessionFromRequestHeaders(headers).orElse(null);
+        if (session == null) {
+            LOGGER.error("Session cannot be found");
+            return Optional.of(generateApiGatewayProxyErrorResponse(400, ERROR_1000));
+        } else {
+            LOGGER.info("{} processing request for session {}", getClass(), session.getSessionId());
+        }
+
+        clientSession =
+                clientSessionService.getClientSessionFromRequestHeaders(headers).orElse(null);
+        if (clientSession == null) {
+            LOGGER.error(
+                    "{} ClientSession not found for session {}",
+                    getClass(),
+                    session.getSessionId());
+            return Optional.of(generateApiGatewayProxyErrorResponse(400, ERROR_1018));
+        }
+
+        clientRegistry = getClientRegistryForClientSession(clientSession);
+        if (clientRegistry == null) {
+            LOGGER.error(
+                    "{} ClientRegistry not found for session {}",
+                    getClass(),
+                    session.getSessionId());
+            return Optional.of(generateApiGatewayProxyErrorResponse(400, ERROR_1015));
+        }
+        return Optional.empty();
+    }
+
+    private ClientRegistry getClientRegistryForClientSession(ClientSession clientSession) {
+        return clientSessionService
+                .getClientIdForClientSession(clientSession)
+                .map(
+                        id -> {
+                            return clientService.getClient(id).orElse(null);
+                        })
+                .orElse(null);
+    }
+
+    protected boolean isTestClientSession(ClientSession clientSession) {
+        return clientSessionService
+                .getClientIdForClientSession(clientSession)
+                .map(
```

Comment:
Can we use a `flatMap(clientService::getClient)` here instead to collapse the inner Optional?

Reply:
updated